File: java/arcs/core/storage/referencemode/ReferenceModeStorageKey.kt
/*
 * Copyright 2020 Google LLC.
 *
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 *
 * Code distributed by Google as part of this project is also subject to an additional IP rights
 * grant found at
 * http://polymer.github.io/PATENTS.txt
 */

package arcs.core.storage.referencemode

import arcs.core.storage.StorageKey
import arcs.core.storage.StorageKeySpec
import arcs.core.storage.StorageKeyUtils
import arcs.core.storage.embed

/**
 * Special subclass of [StorageKey] used to identify data managed by a
 * [arcs.storage.ReferenceModeStore].
 */
data class ReferenceModeStorageKey(
  val backingKey: StorageKey,
  val storageKey: StorageKey
) : StorageKey(protocol) {

  init {
    // This is a overly strict check as some combinations of different protocols are fine, so it
    // can be relaxed if needed (see [StorageAdapter] for a more precise check).
    require(backingKey.protocol == storageKey.protocol) {
      "Different protocols (${backingKey.protocol} and ${storageKey.protocol}) in a " +
        "ReferenceModeStorageKey can cause problems with garbage collection if the " +
        "backing key is in the database and the container key isn't."
    }
  }

  override fun childKeyWithComponent(component: String): StorageKey =
    ReferenceModeStorageKey(backingKey, storageKey.childKeyWithComponent(component))

  override fun toKeyString(): String = "{${backingKey.embed()}}{${storageKey.embed()}}"

  override fun toString(): String = super.toString()

  companion object : StorageKeySpec<ReferenceModeStorageKey> {
    override val protocol = "reference-mode"
    override fun parse(rawKeyString: String): ReferenceModeStorageKey {
      val invalidFormatMessage: () -> String =
        { "Invalid format for ReferenceModeStorageKey" }
      val storageKeys = StorageKeyUtils.extractKeysFromString(rawKeyString)
      require(storageKeys.size == 2, invalidFormatMessage)

      return ReferenceModeStorageKey(
        storageKeys[0],
        storageKeys[1]
      )
    }
  }
}

File: src/runtime/storageNG/reference-mode-storage-key.ts
/**
 * @license
 * Copyright (c) 2019 Google Inc. All rights reserved.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * Code distributed by Google as part of this project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */

import {StorageKey} from './storage-key.js';

export class ReferenceModeStorageKey extends StorageKey {
  public static readonly protocol = 'reference-mode';
  constructor(public backingKey: StorageKey, public storageKey: StorageKey) {
    super(ReferenceModeStorageKey.protocol);
  }

  toString(): string {
    return `${this.protocol}://{${this.backingKey.embedKey()}}{${this.storageKey.embedKey()}}`;
  }

  childWithComponent(component: string): StorageKey {
    return new ReferenceModeStorageKey(this.backingKey, this.storageKey.childWithComponent(component));
  }

  static fromString(key: string, parse: (key: string) => StorageKey): ReferenceModeStorageKey {
    const match = key.match(/^reference-mode:\/\/{((?:\}\}|[^}])+)}{((?:\}\}|[^}])+)}$/);

    if (!match) {
      throw new Error(`Not a valid ReferenceModeStorageKey: ${key}.`);
    }
    const [_, backingKey, storageKey] = match;

    return new ReferenceModeStorageKey(
      parse(StorageKey.unembedKey(backingKey)),
      parse(StorageKey.unembedKey(storageKey))
    );
  }
}
